flink-state-backends/flink-statebackend-heap-spillable/src/main/java/org/apache/flink/runtime/state/heap/CopyOnWriteSkipListStateMap.java

```diff
@@ -0,0 +1,1527 @@
+
+			currentLevel--;
+			if (currentLevel < 0) {
+				break;
+			}
+			currentNode = helpGetNextNode(prevNode, currentLevel);
+		}
+
+		// find the first node that has not been logically removed
+		while (currentNode != NIL_NODE) {
+			if (isNodeRemoved(currentNode)) {
+				currentNode = helpGetNextNode(currentNode, 0);
+				continue;
+			}
+
+			c = compareNamespaceAndNode(namespaceByteBuffer, namespaceOffset, namespaceLen, currentNode);
+			if (c == 0) {
+				return currentNode;
+			}
+
+			if (c < 0) {
+				break;
+			}
+		}
+
+		return NIL_NODE;
+	}
+
+	/**
+	 * Try to delete some nodes that has been logically removed.
+	 */
+	private void tryToDeleteNodesPhysically() {
+		if (highestRequiredSnapshotVersion != 0) {
+			return;
+		}
+
+		int threshold = (int) (totalSize * logicalRemovedKeysRatio);
+		int size = logicallyRemovedNodes.size();
+		if (size > threshold) {
+			deleteLogicallyRemovedNodes(size - threshold);
+		}
+	}
+
+	private void deleteLogicallyRemovedNodes(int maxNodes) {
+		int count = 0;
+		Iterator<Long> nodeIterator = logicallyRemovedNodes.iterator();
+		while (count < maxNodes && nodeIterator.hasNext()) {
+			deleteNode(nodeIterator.next());
+			nodeIterator.remove();
+			totalSize--;
+			count++;
+		}
+	}
+
+	private void deleteNode(long node) {
+		long prevNode = SkipListUtils.findPredecessor(node, 1, levelIndexHeader, spaceAllocator);
+		long currentNode = helpGetNextNode(prevNode, 0);
+		while (currentNode != node) {
+			prevNode = currentNode;
+			currentNode = helpGetNextNode(prevNode, 0);
+		}
+
+		long nextNode = helpGetNextNode(currentNode, 0);
+		doPhysicalRemove(currentNode, prevNode, nextNode);
+	}
+
+	/**
+	 * Release all resource used by the map.
+	 */
+	private void releaseAllResource() {
+		long node = levelIndexHeader.getNextNode(0);
+		while (node != NIL_NODE) {
+			long nextNode = helpGetNextNode(node, 0);
+			long valuePointer = SkipListUtils.helpGetValuePointer(node, spaceAllocator);
+			spaceAllocator.free(node);
+			SkipListUtils.removeAllValues(valuePointer, spaceAllocator);
+			node = nextNode;
+		}
+		totalSize = 0;
+		logicallyRemovedNodes.clear();
+	}
+
+	/**
+	 * Returns the value pointer used by the snapshot of the given version.
+	 *
+	 * @param snapshotVersion version of snapshot.
+	 * @return the value pointer used by the given snapshot. NIL_VALUE_POINTER
+	 * 	will be returned if there is no value for this snapshot.
+	 */
+	long getValueForSnapshot(long node, int snapshotVersion) {
+		long snapshotValuePointer = NIL_VALUE_POINTER;
+		long valuePointer = SkipListUtils.helpGetValuePointer(node, spaceAllocator);
+
+		while (valuePointer != NIL_VALUE_POINTER) {
+			int version = SkipListUtils.helpGetValueVersion(valuePointer, spaceAllocator);
+
+			// the first value whose version is less than snapshotVersion
+			if (version < snapshotVersion) {
+				snapshotValuePointer = valuePointer;
+				break;
+			}
+
+			valuePointer = SkipListUtils.helpGetNextValuePointer(valuePointer, spaceAllocator);
```

Comment:
Since the next value is an `int`, using iterator will introduce additional auto-boxing cost, but let me make the change due to state access is single-threaded thus auto-boxing won't cause big issue.